runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -9086,6 +9086,150 @@ inlineNanoTime(
 #endif
 #endif // LINUX
 
+TR::Register* J9::X86::TreeEvaluator::inlineMathFma(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild = node->getThirdChild();
+
+   TR::Register *lhsReg = NULL;
+   TR::Register *midReg = NULL;
+   TR::Register *rhsReg = NULL;
+   TR::Register *result = cg->allocateRegister(TR_FPR);
+
+   bool memLoadLhs = !firstChild->getRegister() && firstChild->getReferenceCount() == 1
+                     && firstChild->getOpCode().isLoadVar();
+
+   bool memLoadMiddle = !secondChild->getRegister() && secondChild->getReferenceCount() == 1
+                        && secondChild->getOpCode().isLoadVar();
+
+   bool memLoadRhs = !thirdChild->getRegister() && thirdChild->getReferenceCount() == 1
+                     && thirdChild->getOpCode().isLoadVar();
+
+   bool is64Bit = node->getDataType().isDouble();
+
+   TR::InstOpCode::Mnemonic fpMovRegRegOpcode = is64Bit ? TR::InstOpCode::MOVSDRegReg : TR::InstOpCode::MOVSSRegReg;
+   result->setIsSinglePrecision(!is64Bit);
+
+   TR_ASSERT_FATAL(cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_FMA), "Cannot generate inline fma implementation without FMA extensions");
+
+   // Choose fma instruction carefully, based on operand form, to reduce number of copies
+   if (memLoadLhs)
+      {
+      TR::InstOpCode::Mnemonic opcode = is64Bit ? TR::InstOpCode::VFMADD231SDRegRegMem : TR::InstOpCode::VFMADD231SSRegRegMem;
+      TR::MemoryReference *lhsMR = generateX86MemoryReference(firstChild, cg);
+
+      if (memLoadRhs)
+         {
+         // a (2) * b (3) + c (1)
+         TR::MemoryReference *rhsMR = generateX86MemoryReference(thirdChild, cg);
+         generateRegMemInstruction(TR::InstOpCode::MOVSRegMem(is64Bit), node, result, rhsMR, cg);
+
+         midReg = cg->evaluate(secondChild);
+         memLoadMiddle = false; // No choice but to evaluate;
+         generateRegRegMemInstruction(opcode, node, result, midReg, lhsMR, cg);
+         }
+      else if (memLoadMiddle)
+         {
+         // fma = a (1) * b (3) + c (2)
+         opcode = is64Bit ? TR::InstOpCode::VFMADD132SDRegRegMem : TR::InstOpCode::VFMADD132SSRegRegMem;
+
+         TR::MemoryReference *midMR = generateX86MemoryReference(secondChild, cg);
+         rhsReg = cg->evaluate(thirdChild);
+
+         generateRegMemInstruction(TR::InstOpCode::MOVSRegMem(is64Bit), node, result, lhsMR, cg);
+         generateRegRegMemInstruction(opcode, node, result, rhsReg, midMR, cg);
+         }
+      else
+         {
+         // fma = a (2) * b (1) + c (3)
```

Comment:
I think this should be

```suggestion
         // fma = a (2) * b (3) + c (1)
```

Reply:
done